app/routes/api+/custom-fields.bulk-actions.ts

```diff
@@ -0,0 +1,158 @@
+import {
+  bulkActivateOrDeactivateCustomFields,
+  countActiveCustomFields,
+} from "~/modules/custom-field/service.server";
+import { getOrganizationTierLimit } from "~/modules/tier/service.server";
+import { checkExhaustiveSwitch } from "~/utils/check-exhaustive-switch";
+import { sendNotification } from "~/utils/emitter/send-notification.server";
+import { makeShelfError, ShelfError } from "~/utils/error";
+import { assertIsPost, data, error, parseData } from "~/utils/http.server";
+import { ALL_SELECTED_KEY } from "~/utils/list";
+import {
+  PermissionAction,
+  PermissionEntity,
+} from "~/utils/permissions/permission.validator.server";
+import { requirePermission } from "~/utils/roles.server";
+import {
+  canCreateMoreCustomFields,
+  willExceedCustomFieldLimit,
+} from "~/utils/subscription.server";
+
+export async function action({ request, context }: ActionFunctionArgs) {
+  const authSession = context.getSession();
+  const { userId } = authSession;
+
+  try {
+    assertIsPost(request);
+
+    const formData = await request.formData();
+
+    const { intent } = parseData(
+      formData,
+      z.object({ intent: z.enum(["bulk-activate", "bulk-deactivate"]) })
+    );
+
+    const intentToActionMap: Record<typeof intent, PermissionAction> = {
+      "bulk-activate": PermissionAction.update,
+      "bulk-deactivate": PermissionAction.update,
+    };
+
+    const { organizationId, organizations } = await requirePermission({
+      userId,
+      request,
+      entity: PermissionEntity.customField,
+      action: intentToActionMap[intent],
+    });
+
+    switch (intent) {
+      case "bulk-activate": {
+        const { customFieldIds } = parseData(
+          formData,
+          BulkActivateCustomFieldSchema
+        );
+
+        const newActivatingFields = await db.customField.count({
+          where: customFieldIds.includes(ALL_SELECTED_KEY)
+            ? { organizationId }
+            : { id: { in: customFieldIds } },
+        });
+
+        /** We have to make sure that user is not already at the limit */
+        const tierLimit = await getOrganizationTierLimit({
+          organizationId,
+          organizations,
+        });
+
+        const totalActiveCustomFields = await countActiveCustomFields({
+          organizationId,
+        });
+
+        const canCreateMore = canCreateMoreCustomFields({
+          tierLimit,
+          totalCustomFields: totalActiveCustomFields,
+        });
+
+        if (!canCreateMore) {
+          throw new ShelfError({
+            cause: null,
+            message:
+              "You have reached your limit of active custom fields. Please upgrade your plan to add more.",
+            additionalData: { userId, totalActiveCustomFields, tierLimit },
+            label: "Custom fields",
+            status: 403,
+            shouldBeCaptured: false,
+          });
+        }
+
+        /** We have to make sure that the new activating fields is not exceeding the allowed limit */
```

Comment:
This is kind of the same as above. Considering this is being called inside the action, you should rather make an assert function that directly throws the error and use it here.